File: 241-different-ways-to-add-parentheses/different-ways-to-add-parentheses.cpp
class Solution {
public:
    vector<int> diffWaysToCompute(string expression) {
        vector<int> result;
        for (int i = 0; i < expression.size(); ++i) {
            char oper = expression[i];
            if (oper == '+' || oper == '-' || oper == '*') {
                vector<int> s1 = diffWaysToCompute(expression.substr(0, i));
                vector<int> s2 = diffWaysToCompute(expression.substr(i + 1));
                for (int a : s1) {
                    for (int b : s2) {
                        if (oper == '+') 
                            result.push_back(a + b);
                        else if (oper == '-') 
                            result.push_back(a - b);
                        else if (oper == '*') 
                            result.push_back(a * b);
                    }
                }
            }
        }
        if (result.empty()) 
            result.push_back(stoi(expression));
        return result;
    }
};

File: 241-different-ways-to-add-parentheses/different-ways-to-add-parentheses.py
class Solution:
    def diffWaysToCompute(self, expression: str) -> List[int]:
        result = []
        for i in range(len(expression)):
            oper = expression[i]
            if oper == "+" or oper == "-" or oper == "*":
                sub_str1 = expression[0 : i]
                sub_str2 = expression[i + 1 : ]
                s1 = self.diffWaysToCompute(sub_str1)
                s2 = self.diffWaysToCompute(sub_str2)
                for i in s1:
                    for j in s2:
                        if oper == "*":
                            result.append(int(i) * int(j))
                        if oper == "+":
                            result.append(int(i) + int(j))
                        if oper == "-":
                            result.append(int(i) - int(j))
        if len(result) == 0:
            result.append(int(expression))
        return result